src/controllers/auth.js

```diff
@@ -1,5 +1,217 @@
+const { OAuth2Client } = require('google-auth-library');
+const { StudentModel } = require('../models/student');
+
+const client = new OAuth2Client(process.env.GOOGLE_CLIENT_ID);
+/**
+ * Note: for google authentication:
+ * 1. add your google client_id to env file.
+ * 2. add your google client_id to ejs /login index.js file.
+ */
+
+/**
+ * Returns jwt token
+ * @returns {string}
+ */
+function generateAccessToken(student) {
+  return jwt.sign(student, process.env.TOKEN_SECRET, {
+    expiresIn: process.env.TOKEN_EXPIRES,
+  });
+}
+
+/**
+ * Generates and return unique username
+ * @returns {string}
+ */
+async function getUniqueUserName(proposedName) {
+  const username = slugify(proposedName, {
+    replacement: '_', // replace spaces with replacement character, defaults to `-`
+    lower: true, // convert to lower case, defaults to `false`
+    locale: 'tr', // language code of the locale to use
+  });
+  const student = await StudentModel.findOne({ username });
+  if (student) {
+    // if student exist call function again with adding random numbers
+    const newProposedName = proposedName + Math.floor(Math.random() * 100 + 1);
+    return getUniqueUserName(newProposedName);
+  }
+  // if student not exist return username
+  return username;
+}
+
 module.exports = {
-  signin: async () => {},
-  signInWithGmail: async () => {},
-  signup: async () => {},
+  signin: async (req, res) => {
+    const { email, password } = req.body;
+
+    // check username and password
+    if (!email && !password) {
+      return res
+        .status(400)
+        .json({ error: 'email and password fields are required' });
+    }
+
+    try {
+      // find student from email
+      const student = await StudentModel.findOne({ email });
+
+      if (!student) {
+        return res.status(400).json({ error: 'wrong email or password' });
+      }
+      if (student.isSignUpWithGoogle) {
+        // student should login via using gmail.
+        // because there is not any password in database for this student.
+        // it requires google oauth verification.
+        return res.status(400).json({
+          error: 'This account is linked to Google. Please log in via Google',
+        });
+      }
+
+      // compare sent password to saved one
+      const result = await bcrypt.compare(password, student.passwordHash);
+      if (!result) {
+        return res.status(400).json({ error: 'wrong email or password' });
+      }
+
+      // generate jwt token and send it to client with student information in payload
+      const token = generateAccessToken({ student });
+      return res.status(200).json(token);
+    } catch (err) {
+      return res.status(400).json(err);
+    }
+  },
+  signup: async (req, res) => {
+    const { firstName, lastName, password, avatar, email, isTutor, subjects } =
+      req.body;
+
+    // password are required,
+    // length should longer than 6 characters
+    if (!password || password.length < 6) {
```

Comment:
for extra security you can make it 8 character at least

Reply:
Thank you I did it.